pkg/backup/item_collector.go

```diff
@@ -296,50 +296,70 @@ func (r *itemCollector) getResourceItems(log logrus.FieldLogger, gv schema.Group
 			labelSelector = metav1.FormatLabelSelector(selector)
 		}
 
-		log.Info("Listing items")
-		unstructuredItems := make([]unstructured.Unstructured, 0)
-
-		if r.pageSize > 0 {
-			// If limit is positive, use a pager to split list over multiple requests
-			// Use Velero's dynamic list function instead of the default
-			listPager := pager.New(pager.SimplePageFunc(func(opts metav1.ListOptions) (runtime.Object, error) {
-				return resourceClient.List(opts)
-			}))
-			// Use the page size defined in the server config
-			// TODO allow configuration of page buffer size
-			listPager.PageSize = int64(r.pageSize)
-			// Add each item to temporary slice
-			list, paginated, err := listPager.List(context.Background(), metav1.ListOptions{LabelSelector: labelSelector})
-			if err != nil {
-				log.WithError(errors.WithStack(err)).Error("Error listing resources")
-				continue
-			}
-			if !paginated {
-				log.Infof("list for groupResource %s was not paginated", gr)
-			}
-			err = meta.EachListItem(list, func(object runtime.Object) error {
-				u, ok := object.(*unstructured.Unstructured)
-				if !ok {
-					log.WithError(errors.WithStack(fmt.Errorf("expected *unstructured.Unstructured but got %T", u))).Error("unable to understand entry in the list")
-					return fmt.Errorf("expected *unstructured.Unstructured but got %T", u)
-				}
-				unstructuredItems = append(unstructuredItems, *u)
-				return nil
-			})
-			if err != nil {
-				log.WithError(errors.WithStack(err)).Error("unable to understand paginated list")
-				continue
+		var labelSelectors []string
+		if r.backupRequest.Spec.LabelSelectors != nil {
+			for _, s := range r.backupRequest.Spec.LabelSelectors {
+				labelSelectors = append(labelSelectors, metav1.FormatLabelSelector(s))
 			}
 		} else {
-			// If limit is not positive, do not use paging. Instead, request all items at once
-			unstructuredList, err := resourceClient.List(metav1.ListOptions{LabelSelector: labelSelector})
+			labelSelectors = []string{}
+		}
+
+		log.Info("Listing items")
+		unstructuredItems := make([]unstructured.Unstructured, 0)
+		for _, label := range labelSelectors {
+			unstructuredList, err := resourceClient.List(metav1.ListOptions{LabelSelector: label})
 			if err != nil {
 				log.WithError(errors.WithStack(err)).Error("Error listing items")
 				continue
 			}
 			unstructuredItems = append(unstructuredItems, unstructuredList.Items...)
 		}
 
+		// use labelSelector (singular) behavior only when the user has not specified labelSelectors(plural) in backup API
+		if len(labelSelectors) == 0 {
```

Comment:
here when len(labelSelectors) >0, the labelSelector (singular) is ignored, is it the design?

Reply:
Hmm. I may be misremembering, but I thought that if both single and plural were supplied, both would have to match.

Reply:
Yes @reasonerjt @sseago it would be by design. `OrLabelSelectors` should take precedence, because:
- first of all specifying `LabelSelector` and `orLabelSelectors` would not be a real world use-case
- secondly, the AND of labels is already sufficed by `LabelSelector` and it would not make sense to take another AND of `LabelSelector` and `OrLabelSelectors`. I believe the AND behavior would definitely confuse the user
- Finally, this preserves the current velero behavior where the `OrLabelSelectors` does not exist, processing based on the `LabelSelector` only when the `OrLabelSelectors` is empty helps reinforce this.

Reply:
@shubham-pampattiwar So yes, this is an edge case that shouldn't happen under normal circumstances, but ignoring `LabelSelector` when the other is present is inconsistent with existing behavior. Currently, if a field is present, its requirements must be met. If we don't intend to enforce the requirements of both when both are present, then instead of ignoring `LabelSelector` we should raise an error if both are set. Otherwise,  a user might expect the field to be enforced.

Reply:
Thanks for the suggestions guys. I have kept the processing of `LabelSelectors` and `OrLabeleSelectors` separate from each other. I have added validations based on the existence of both the spec, if the user specifies both `LabelSelectors` and `OrLabelSelectors` the backup/restore will have an error and a `FailedValidation` phase, indicating to the user that both spec fields cannot be used at the same time.